### ISSUE TITLE
False positive for declarationless rules in no-duplicate-selectors

### ISSUE THREAD

**Issue by subzey** [stylelint/stylelint#2198]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A feature request. Add a secondary option `ignoreEmpty` to `no-duplicate-selectors`, so blocks with no content won't count.

The use case:

```less
.some-block {
  & {
    color: black;
  }

  &--error {
    color: red;
  }
}
```

This LESS code is linted as 

```css
.some-block {}
.some-block {
  color: black;
}
.some-block--error {
  color: red;
}
```
triggering a `no-duplicate-selectors` error/warning. `ignoreEmpty` would allow using such a style, yet catching the "real" errors, like this:

```less
.some-block {
  color: black;

  & { // ERROR!
    color: green;
  }
}
```

> Which rule, if any, is this issue related to?

`no-duplicate-selectors`

> What CSS is needed to reproduce this issue?

e.g.

```less
.some-block {
  & {
    color: black;
  }

  &--error {
    color: red;
  }
}
```
> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
  "rules": {
    "no-duplicate-selectors": true
  }
}
```

> Which version of stylelint are you using?

`7.6.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI, as a git hook and with SublimeLinter-contrib-stylelint

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's about LESS nesting

> What did you expect to happen?

`&` within a rule without properties is not treated as a duplicate block

> What actually happened (e.g. what warnings or errors you are getting)?

`&` within a rule without properties is treated as a duplicate block

**Comment by evilebottnawi** [comment]
@subzey I think it is bug, not feature, we should ignore `selectors` which contains only nesting

**Comment by evilebottnawi** [comment]
@subzey it should be easy fix, add `isContainsOnlyNesting` utils (maybe best name) and skip this selectors.

**Comment by jeddy3** [comment]
> it should be easy fix, add isContainsOnlyNesting utils (maybe best name) and skip this selectors.

SGTM. Let's go with`hasOnlyNestedRules` to be in line with our other `has*` utils like `hasEmptyBlock`.

**Comment by hudochenkov** [comment]
This rule fails if a CSS rule has a nested rule with selector `&`. I think #2199 fails because of this too.

When I started to figure out a solution for this problem and edge cases, I wrote following text based on the discussion above:

---

What `hasOnlyNestedRules` would check? And where it will be applied?

This shouldn't show warnings but shows now:

```scss
.foo {
	& {}
}
.foo {
	&--bar {}
	& {}
}
```

This should show warnings, and it's ok:

```scss
.foo {
	color: red;
	& {
		color: blue;
	}
}
```

How about this:

```scss
.foo {
	/* comment 1 */
	& {}

	/* comment 2 */
	&--bar {}
}
```

Should it fails, if there are nested rules and comments? Processed CSS would have two `.foo` rule-sets: one with comments only, and one with content of `& {}`.

There are few more edge cases for non-standard syntax:

```scss
.foo {
	@include icon;

	& {
		color: red;
	}
}
```

```scss
.foo {
	@at-root & {
		color: red;
	}
}
```

```css
.foo {
	@media (min-width: 300px) {
		& {
			color: pink;
		}
	}
}
```

---

Now I re-read `no-duplicate-selectors` description and where it applies. I believe we were thinking in a wrong direction. The rule is about selectors only, not about rules content or how preprocessors will compile the code.

We shouldn't check rule content because we can't predict what preprocessors will compile.

Following code:

```scss
.foo {
	& {}
}
```

Can be compiled as `.foo {}` or as `.foo {} .foo {}` depending on preprocessor and its settings.

We don't know how it will compile. But in most cases, it will compile as `.foo {} .foo {}`, because usually there is declarations or other content in parent rule and nested rules.

To sum up, I think it's a correct behavior for `no-duplicate-selectors` to show warnings for this code:

```scss
.foo {
	& {}
}
```

**Comment by evilebottnawi** [comment]
@hudochenkov can you send failed test PR? so it will be much easier to solve the problem

**Comment by hudochenkov** [comment]
@evilebottnawi how failed test should be in a test code? Should it really fail tests?

I believe this rule should throw warnings in this particular case, so I can add tests in `reject` array.

**Comment by jeddy3** [comment]
@hudochenkov Thanks for doing the investigation.

> To sum up, I think it's a correct behavior for no-duplicate-selectors to show warnings for this code:

I agree with your assessment. Does that means this issue can be closed?

**Comment by hudochenkov** [comment]
Yeah, after I'll add tests, so we won't forget about correct behaviour :)

**Comment by hudochenkov** [comment]
Closed because it works as expected. Explanation in a [comment](https://github.com/stylelint/stylelint/issues/2198#issuecomment-274276106) above.

**Comment by 7iomka** [comment]
Hi. How to make exception for this classic situation? On line 22 I have an error 
> Unexpected duplicate selector ".how-steps-waves::before", first used at line 12 (no-duplicate-selectors)stylelint(no-duplicate-selectors)
![image](https://user-images.githubusercontent.com/18473137/55881282-04d34280-5bab-11e9-8498-49b6059fbe07.png)

**Comment by hudochenkov** [comment]
@7iomka you have duplicate selector on line 13.

**Comment by LuXDAmore** [comment]
@hudochenkov , What do you think about this `.scss` file?
 
```css

  .class {
  
    &, /* <-- no-duplicate-selectors error here */
    > .content {
  
         display: block;
  
    }
  
  }

```

The error still in `&,` but i think should be something allowed, because i prefer to not duplicate code in that case.

**Comment by rmachado-studocu** [comment]
I'm guessing I'll have to disable this rule, with my SCSS files, given this: https://sass-lang.com/documentation/breaking-changes/mixed-decls/
